From 44a62980d836d3e2b8e66244ef1fc3f72aac57b1 Mon Sep 17 00:00:00 2001
From: Tom Cosgrove <tom.cosgrove@arm.com>
Date: Wed, 25 Jan 2023 19:34:25 +0000
Subject: [PATCH 1/2] Enable AES optimisation on Apple Silicon M2-based systems

Gives a performance enhancement of 16-38%, similar to the M1.

Reviewed-by: Tomas Mraz <tomas@openssl.org>
Reviewed-by: Hugo Landau <hlandau@openssl.org>
Reviewed-by: Paul Dale <pauli@openssl.org>
(Merged from https://github.com/openssl/openssl/pull/20141)

(cherry picked from commit d79bb5316e1318bd776d6b2d6723a36778e07f9d)

Upstream-Status: Backport [d79bb5316e1318bd776d6b2d6723a36778e07f9d]
---
 crypto/armcap.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/crypto/armcap.c b/crypto/armcap.c
index 71296786c3..68f735baec 100644
--- ./crypto/armcap.c
+++ ./crypto/armcap.c
@@ -254,7 +254,8 @@ void OPENSSL_cpuid_setup(void)
             OPENSSL_armcap_P |= ARMV8_SHA3;
             len = sizeof(uarch);
             if ((sysctlbyname("machdep.cpu.brand_string", uarch, &len, NULL, 0) == 0) &&
-                (strncmp(uarch, "Apple M1", 8) == 0))
+               ((strncmp(uarch, "Apple M1", 8) == 0) ||
+                (strncmp(uarch, "Apple M2", 8) == 0)))
                 OPENSSL_armcap_P |= ARMV8_UNROLL8_EOR3;
         }
     }
-- 
2.40.0

